server/common/utils/userUtils.js

```diff
@@ -125,11 +128,154 @@ const responseWithUserData = user => {
   return data;
 };
 
+const checkIfUserIsTheOnlyOwner = async (listModel, cohortModel, userId) => {
+  const errors = {};
+
+  const lists = await listModel
+    .find(
+      {
+        $and: [{ ownerIds: userId }, { ownerIds: { $size: 1 } }],
+        'viewersIds.1': { $exists: true }
+      },
+      'name'
+    )
+    .lean()
+    .exec();
+
+  const cohorts = await cohortModel
+    .find(
+      {
+        $and: [{ ownerIds: userId }, { ownerIds: { $size: 1 } }],
+        'memberIds.1': { $exists: true }
+      },
+      'name'
+    )
+    .lean()
+    .exec();
+
+  if (lists.length > 0) {
+    errors.lists = lists;
+  }
+
+  if (cohorts.length > 0) {
+    errors.cohorts = cohorts;
+  }
+
+  return errors;
+};
+
+const deleteAccountDetails = async (userModel, user) => {
+  const { _id: userId, createdAt, email } = user;
+  const hashedEmail = bcrypt.hashSync(email, 12);
+  const removedUser = new User({
+    _id: userId,
+    createdAt,
+    email: hashedEmail,
+    updatedAt: new Date()
+  });
+
+  await userModel.deleteOne({ email });
+  await removedUser.save();
+};
+
+const deleteUserLists = async (listModel, userId) =>
+  listModel
+    .updateMany(
+      { viewersIds: userId, 'viewersIds.1': { $exists: false } },
+      {
+        isArchived: true,
+        isDeleted: true
+      }
+    )
+    .exec();
+
+const removeUserFromLists = io => async (listModel, displayName, userId) => {
+  const lists = await listModel.find({ viewersIds: userId });
+  const listPromises = [];
+
+  lists.forEach(async list => {
```

Comment:
Flatmap this shit - `Promise.all(lists.flatMap(list => ... [list.save(), leaveList(io)(data)]))`

Reply:
we have node 10. and flatMap was added to node 11

Reply:
Catch some ponyfill then and consider using it -> `const flatMap = f => xs => xs.reduce((ys, y) => ys.concat(f(y)), []);` (included in std https://github.com/sandstreamdev/std/blob/master/array/flatMap.js)